src/common/download.cpp

```diff
@@ -40,27 +41,86 @@
 
 namespace tools
 {
-  static bool download_thread(const std::string &path, const std::string &url)
+  struct download_thread_control
   {
+    const std::string path;
+    const std::string uri;
+    std::function<void(const std::string&, const std::string&, bool)> result_cb;
+    std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress_cb;
+    bool stop;
+    bool stopped;
+    bool success;
+    std::shared_ptr<boost::thread> thread;
+    boost::mutex mutex;
+
+    download_thread_control(const std::string &path, const std::string &uri, std::function<void(const std::string&, const std::string&, bool)> result_cb, std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress_cb):
+        path(path), uri(uri), result_cb(result_cb), progress_cb(progress_cb), stop(false), stopped(false), success(false) {}
+    void lock() { mutex.lock(); }
+    void unlock() { mutex.unlock(); }
+  };
+
+  static void download_thread(download_thread_control *control)
+  {
+    static std::atomic<unsigned int> thread_id(0);
+
+    MLOG_SET_THREAD_NAME("DL" + std::to_string(thread_id++));
+
+    epee::misc_utils::auto_scope_leave_caller scope_exit_handler = epee::misc_utils::create_scope_leave_handler([&control](){ delete control; });
     try
     {
-      MINFO("Downloading " << url << " to " << path);
+      control->lock();
+      MINFO("Downloading " << control->uri << " to " << control->path);
       std::ofstream f;
-      f.open(path, std::ios_base::binary | std::ios_base::out | std::ios_base::trunc);
+      f.open(control->path, std::ios_base::binary | std::ios_base::out | std::ios_base::trunc);
       if (!f.good()) {
-        MERROR("Failed to open file " << path);
-        return false;
+        MERROR("Failed to open file " << control->path);
+        control->result_cb(control->path, control->uri, control->success);
+        control->stopped = true;
+        control->unlock();
+        return;
       }
       class download_client: public epee::net_utils::http::http_simple_client
       {
       public:
-        download_client(std::ofstream &f): f(f) {}
+        download_client(download_thread_control *control, std::ofstream &f):
+          control(control), f(f), content_length(-1), total(0) {}
         virtual ~download_client() { f.close(); }
+        virtual bool on_header(const epee::net_utils::http::http_response_info &headers)
+        {
+          ssize_t length;
```

Comment:
Why `ssize_t` on all of these calls? IIRC, passing an unsigned type to `get_xtype_from_string` will return false if `m_content_length` is a negative value.